src/Security/Authentication/OAuth/src/JsonKeyClaimAction.cs

```diff
@@ -30,20 +30,27 @@ public JsonKeyClaimAction(string claimType, string valueType, string jsonKey)
         public string JsonKey { get; }
 
         /// <inheritdoc />
-        public override void Run(JObject userData, ClaimsIdentity identity, string issuer)
+        public override void Run(JsonDocument userData, ClaimsIdentity identity, string issuer)
         {
-            var value = userData?[JsonKey];
-            if (value is JValue)
+            if (userData == null || !userData.RootElement.TryGetProperty(JsonKey, out var value))
             {
-                AddClaim(value?.ToString(), identity, issuer);
+                return;
             }
-            else if (value is JArray)
+            if (value.Type == JsonValueType.Array)
             {
-                foreach (var v in value)
+                foreach (var v in value.EnumerateArray())
                 {
-                    AddClaim(v?.ToString(), identity, issuer);
+                    AddClaim(v.ToString(), identity, issuer);
```

Comment:
What happens if a value in the array is a null token? I think add a test

Reply:
Added a test.